addons/mail/controllers/mail.py

```diff
@@ -121,22 +121,18 @@ def _redirect_to_record(cls, model, res_id, access_token=None, **kwargs):
         elif not record_action['type'] == 'ir.actions.act_window':
             return cls._redirect_to_messaging()
 
-        url_params = {
-            'model': model,
-            'id': res_id,
-            'active_id': res_id,
-            'action': record_action.get('id'),
-        }
+        url = '/odoo/%s/%s' % (model, res_id)
         menu_id = request.env['ir.ui.menu']._get_best_backend_root_menu_id_for_model(model)
         if menu_id:
-            url_params['menu_id'] = menu_id
+            url = url + '?menu_id=%s' % menu_id
         view_id = record_sudo.get_formview_id()
         if view_id:
-            url_params['view_id'] = view_id
-
+            if menu_id:
+                url = url + '&amp;view_id=%s' % view_id
```

Comment:
Please use tools to build URLs, don't do it manually

Reply:
Hello @tde-banana-odoo 
Previously we were using `url_encode` tool from werkzeug which was taking a `MultiDict` param and returning a `query string` which we were adding in our url like this - `url = '/web?#%s' % url_encode(url_params)` as fragment(part of url after `#`).
But now the url has been modified as per the [task](https://www.odoo.com/odoo/project/133/tasks/3820230?cids=2) specification which will not consist of fragment(part of url after `#`). In this updated url actions, models and ids are directly added to url's path instead of fragment.

So can you please take a look into the task's purpose & specifications and give me suggestions regarding this?
Thanks!!

Reply:
You can use url_encode for `view_id` and `menu_id`